quill-engine/src/main/scala/io/getquill/sql/SqlQuery.scala

```diff
@@ -32,6 +32,13 @@ sealed trait SetOperation
 case object UnionOperation extends SetOperation
 case object UnionAllOperation extends SetOperation
 
+sealed trait DistinctKind
```

Comment:
Might be easier if you added a boolean check for this:
```scala
sealed trait DistinctKind { def isDistinct: Boolean }
case object DistinctKind {
  case object Distinct extends DistinctKind { val isDistinct: Boolean = true }
  case object DistinctOn extends DistinctKind { val isDistinct: Boolean = true }
  case object None extends DistinctKind { val isDistinct: Boolean = false }
}
```
Then you can do `x.distinct.isDistinct` wherever it's checking `x.distinct`.